### DISCUSSION TITLE
Rotate & Moving 3D Model in MapLibre

### DISCUSSION THREAD

**Discussion by Takieddinex** [maplibre/maplibre-gl-js#1548] [Q&A]
I would to know if it is possible to rotate and move a 3D model added on a maplibre map. 
I would like to see your examples.
![image](https://user-images.githubusercontent.com/95845810/165779705-5c628cf7-1537-4894-9710-be219f7881a0.png)

**Comment by wipfli** [comment]
Does https://maplibre.org/maplibre-gl-js-docs/example/add-3d-model/ help you?

**Comment by Takieddinex** [comment]
@wipfli Thank you, I started with this example but now I'm looking at how I combine ThreeJs and MapLibre to rotate and move my 3D model 


I was using this code to make my model move but only according to the two origins variables. I'm trying to make it more precise by navigating with the mouse 



`
var modelOrigin = [4.8499, 45.7770]; 
                    var modelOrigin2 = [4.8519, 45.7756];   
                    var modelAltitude = 0;
                    var modelRotate = [Math.PI / 2, 5.5, 0];


`
`





document.getElementById('move').addEventListener('click', function () {
			// Fly to a random location by offsetting the point -74.50, 40
			// by up to 5 degrees.
			modelAsMercatorCoordinate = maplibregl.MercatorCoordinate.fromLngLat(
				modelOrigin2,
				modelAltitude
			);

			// transformation parameters to position, rotate and scale the 3D model onto the map
			modelTransform = {
				translateX: modelAsMercatorCoordinate.x,
				translateY: modelAsMercatorCoordinate.y,
				translateZ: modelAsMercatorCoordinate.z,
				rotateX: modelRotate[0],
				rotateY: modelRotate[2],
				rotateZ: modelRotate[2],
				/* Since our 3D model is in real world meters, a scale transform needs to be
				* applied since the CustomLayerInterface expects units in MercatorCoordinates.
				*/
				scale: modelAsMercatorCoordinate.meterInMercatorCoordinateUnits()
			};

		});


`

**Comment by timokorkalainen** [comment]
You can use ThreeJS with MapLibre for sure. The way you do it is to place the ThreeJS scene origin using the method shown in the 3D model example and then using ThreeJS's own methods to place the 3D models relatively to the origin. If you need to use coordinates for the models' locations, first perform conversion calculations of the distances between the target locations and the scene origin coordinates and then apply the correct transforms to the models.

**Comment by HarelM** [comment]
Is there an issue here or should this be converted to a discussion?

**Comment by timokorkalainen** [comment]
Discussion would indeed be a better place for this. I do not think there are any issues with this.

**Comment by ivannovazzi** [comment]
For anyone wanting to animate a moving vehicle model in MapLibre with *realistic* movement along actual roads (not just a scripted path), [Moveet](https://github.com/ivannovazzi/moveet) might be a useful data source.

It's an open-source fleet simulator that streams real-time vehicle positions over WebSocket — the vehicles navigate real road networks (OpenStreetMap data) using A* pathfinding, so the coordinates and heading you receive follow actual road geometry. You can pipe those positions into MapLibre to drive your 3D model's location and rotation.

The simulator itself uses this approach for its own D3.js visualization, so the data is designed to be consumed directly by a map renderer.
